File: computing/interview/design-patterns/behavioral/memento/python/memento_demo.py
from __future__ import annotations
from dataclasses import dataclass


@dataclass(frozen=True)
class TextMemento:
    state: str


class TextEditor:
    def __init__(self) -> None:
        self.content = ""

    def type(self, text: str) -> None:
        self.content += text

    def save(self) -> TextMemento:
        return TextMemento(self.content)

    def restore(self, memento: TextMemento) -> None:
        self.content = memento.state


class History:
    def __init__(self) -> None:
        self._snapshots: list[TextMemento] = []

    def push(self, memento: TextMemento) -> None:
        self._snapshots.append(memento)

    def pop(self) -> TextMemento:
        if not self._snapshots:
            raise IndexError("No states saved")
        return self._snapshots.pop()


def main() -> None:
    editor = TextEditor()
    history = History()

    editor.type("Hello")
    history.push(editor.save())

    editor.type(" World")
    print(f"Current: {editor.content}")

    editor.restore(history.pop())
    print(f"After undo: {editor.content}")


if __name__ == "__main__":
    main()

File: computing/interview/design-patterns/behavioral/memento/java/MementoDemo.java
package interview.designpatterns.behavioral.memento;

import java.util.Deque;
import java.util.LinkedList;

final class TextMemento {
    private final String state;

    TextMemento(String state) {
        this.state = state;
    }

    String state() {
        return state;
    }
}

final class TextEditor {
    private final StringBuilder content = new StringBuilder();

    void type(String text) {
        content.append(text);
    }

    TextMemento save() {
        return new TextMemento(content.toString());
    }

    void restore(TextMemento memento) {
        content.setLength(0);
        content.append(memento.state());
    }

    String content() {
        return content.toString();
    }
}

final class History {
    private final Deque<TextMemento> snapshots = new LinkedList<>();

    void push(TextMemento memento) {
        snapshots.push(memento);
    }

    TextMemento pop() {
        if (snapshots.isEmpty()) {
            throw new IllegalStateException("No states saved");
        }
        return snapshots.pop();
    }
}

public final class MementoDemo {
    private MementoDemo() {}

    public static void main(String[] args) {
        TextEditor editor = new TextEditor();
        History history = new History();

        editor.type("Hello");
        history.push(editor.save());

        editor.type(" World");
        System.out.printf("Current: %s%n", editor.content());

        editor.restore(history.pop());
        System.out.printf("After undo: %s%n", editor.content());
    }
}
